dask/bag/core.py

```diff
@@ -1014,7 +1014,12 @@ def to_dataframe(self, columns=None):
         columns : list or pandas.DataFrame, optional
             If a list, provides the desired column names. If a
             ``pandas.DataFrame``, it should mirror the column names and dtypes
-            of the output dataframe. If not provided, these will be computed.
+            of the output dataframe. If not provided or provided as list,
+            these will be computed by taking a single element from the first
+            partition, triggering a potentially expensive call to ``compute``.
+            Providing a list is only useful for selecting subset of columns,
+            to avoid an internal compute call you must provide a
+            ``pandas.DataFrame``.
```

Comment:
Perhaps expand the reason for the dataframe?  Alternatively we could lead with the dataframe as the first and correct choice.

> you must provide an ``pandas.DataFrame`` that matches the intended columns and dtype of the full dataset.

Also, you might want to look at #1422 , which may affect this.